pallets/task-scheduler/src/tasks.rs

```diff
@@ -0,0 +1,85 @@
+pub mod error;
+
+use super::pallet::Config;
+use frame_support::traits::Get;
+use sp_core::offchain::Duration;
+use sp_core::Encode;
+use sp_runtime::offchain::storage_lock::BlockAndTime;
+use sp_runtime::offchain::storage_lock::StorageLock;
+use sp_runtime::traits::BlockNumberProvider;
+use sp_runtime::SaturatedConversion;
+
+#[inline]
+#[allow(dead_code)]
+pub(crate) fn storage_key<Id: Encode>(id: &Id) -> Vec<u8> {
+	const TASK_GUARD: &[u8] = b"task-scheduler/task/guard/";
+	id.using_encoded(|encoded_id| TASK_GUARD.iter().chain(encoded_id).copied().collect())
+}
+
+type Lock<'a, BlockNumberProvider> = StorageLock<'a, BlockAndTime<BlockNumberProvider>>;
+
+#[allow(dead_code)]
+pub(crate) fn task_lock<Runtime: Config>(storage_key: &[u8]) -> Lock<frame_system::Pallet<Runtime>>
+where
+	frame_system::Pallet<Runtime>: BlockNumberProvider,
+{
+	let offset = Runtime::UnverifiedTaskTimeout::get()
+		.saturated_into::<u32>()
+		.saturating_sub(2u32);
+	Lock::<frame_system::Pallet<Runtime>>::with_block_and_time_deadline(
+		storage_key,
+		offset,
+		Duration::from_millis(0),
+	)
+}
+
+use crate::SystemConfig;
+use frame_support::dispatch::Vec;
+
+use alloc::fmt::Debug;
+
+pub trait VerifiableTask<Runtime: SystemConfig> {
+	type Call: Clone;
+	type V: Debug;
+	type P: Debug;
+	fn process(
+		&self,
+		deadline: Runtime::BlockNumber,
+	) -> Result<Self::Call, TaskError<Self::V, Self::P>>;
+}
+
+use error::TaskError;
+//this is the iterface you need to implement for onchain types to become Tasks.
+//this is NOT going to be the Config associated type
+pub trait TaskV2<Runtime: frame_system::Config> {
+	type Call;
+	type V;
+	type P;
+	/// task generates its own id.
+	fn to_id(&self) -> Runtime::Hash;
+	//task will know how to check for persistence
+	fn is_persisted(id: &Runtime::Hash) -> bool;
+	/// generates a call after verification.
+	fn post_verification_call(
```

Comment:
I don't think the concept of "verification" should be within the pallet, as not every task will be a "verification" task.

Reply:
I see that the Error types were changed, but what should we call this function?

Reply:
I have two suggestions: `committing_call`/`persistence_call`